src/Sema/LowerZon.zig

```diff
@@ -0,0 +1,902 @@
+const std = @import("std");
+const Zcu = @import("../Zcu.zig");
+const Sema = @import("../Sema.zig");
+const Air = @import("../Air.zig");
+const InternPool = @import("../InternPool.zig");
+const Type = @import("../Type.zig");
+const Value = @import("../Value.zig");
+const Zir = std.zig.Zir;
+const AstGen = std.zig.AstGen;
+const CompileError = Zcu.CompileError;
+const Ast = std.zig.Ast;
+const Allocator = std.mem.Allocator;
+const assert = std.debug.assert;
+const File = Zcu.File;
+const LazySrcLoc = Zcu.LazySrcLoc;
+const Ref = std.zig.Zir.Inst.Ref;
+const NullTerminatedString = InternPool.NullTerminatedString;
+const NumberLiteralError = std.zig.number_literal.Error;
+const NodeIndex = std.zig.Ast.Node.Index;
+const Zoir = std.zig.Zoir;
+
+const LowerZon = @This();
+
+sema: *Sema,
+file: *File,
+file_index: Zcu.File.Index,
+import_loc: LazySrcLoc,
+block: *Sema.Block,
+base_node_inst: InternPool.TrackedInst.Index.Optional = .none,
+
+/// Lowers the given file as ZON.
+pub fn lower(
+    sema: *Sema,
+    file: *File,
+    file_index: Zcu.File.Index,
+    res_ty: Type,
+    import_loc: LazySrcLoc,
+    block: *Sema.Block,
+) CompileError!InternPool.Index {
+    _ = try file.getZoir(sema.pt.zcu);
+    var lower_zon: LowerZon = .{
+        .sema = sema,
+        .file = file,
+        .file_index = file_index,
+        .import_loc = import_loc,
+        .block = block,
+    };
+
+    try lower_zon.checkType(res_ty);
+
+    return lower_zon.lowerExpr(.root, res_ty);
+}
+
+fn checkType(self: *LowerZon, ty: Type) !void {
+    var visited: std.AutoHashMapUnmanaged(Type, void) = .empty;
+    try self.checkTypeInner(ty, &visited);
+}
+
+fn checkTypeInner(
+    self: *LowerZon,
+    ty: Type,
+    visited: *std.AutoHashMapUnmanaged(Type, void),
+) !void {
+    const pt = self.sema.pt;
+
+    // Check if we're a nested optional before we possibly return early due to recursion
+    if (ty.zigTypeTag(pt.zcu) == .optional) {
+        var curr = ty.optionalChild(pt.zcu);
+        while (true) {
+            switch (curr.zigTypeTag(pt.zcu)) {
+                .optional => return self.failUnsupportedResultType(
+                    "nested optionals not available in ZON",
+                    ty,
+                ),
+                .pointer => {
+                    const ptr_info = curr.ptrInfo(pt.zcu);
+                    if (ptr_info.flags.size != .one) break;
+                    curr = .fromInterned(ptr_info.child);
+                },
+                else => break,
+            }
+        }
+    }
```

Comment:
I think I have an alternative way of doing this which avoids having a second loop; if my thinking works I'll push it before merging.

Reply:
Oh, also, `checkType` leaks `visited` lol

Reply:
(Sorry, I'm just throwing comments in here for you to see)
In the runtime parser, `canParseTypeInner`, the `is_comptime` thing is unnecessary, because the parser already errors if a comptime field is provided!